{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Collecting tiledbsoma\n",
      "  Using cached tiledbsoma-1.15.7.tar.gz (685 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Requirement already satisfied: anndata>=0.10.1 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from tiledbsoma) (0.11.0)\n",
      "Requirement already satisfied: attrs>=22.2 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from tiledbsoma) (24.3.0)\n",
      "Requirement already satisfied: numpy in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from tiledbsoma) (1.26.4)\n",
      "Requirement already satisfied: pandas in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from tiledbsoma) (2.2.2)\n",
      "Requirement already satisfied: pyarrow in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from tiledbsoma) (18.1.0)\n",
      "Requirement already satisfied: scanpy>=1.9.2 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from tiledbsoma) (1.10.4)\n",
      "Requirement already satisfied: scipy in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from tiledbsoma) (1.13.1)\n",
      "Collecting somacore==1.0.26 (from tiledbsoma)\n",
      "  Using cached somacore-1.0.26-py3-none-any.whl.metadata (1.5 kB)\n",
      "Requirement already satisfied: typing-extensions in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from tiledbsoma) (4.12.2)\n",
      "Requirement already satisfied: pyarrow-hotfix in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from somacore==1.0.26->tiledbsoma) (0.6)\n",
      "Collecting shapely (from somacore==1.0.26->tiledbsoma)\n",
      "  Using cached shapely-2.0.7-cp312-cp312-win_amd64.whl.metadata (7.1 kB)\n",
      "Requirement already satisfied: array-api-compat!=1.5,>1.4 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from anndata>=0.10.1->tiledbsoma) (1.10.0)\n",
      "Requirement already satisfied: h5py>=3.6 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from anndata>=0.10.1->tiledbsoma) (3.12.1)\n",
      "Requirement already satisfied: natsort in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from anndata>=0.10.1->tiledbsoma) (8.4.0)\n",
      "Requirement already satisfied: packaging>=20.0 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from anndata>=0.10.1->tiledbsoma) (24.2)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from pandas->tiledbsoma) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from pandas->tiledbsoma) (2024.2)\n",
      "Requirement already satisfied: tzdata>=2022.7 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from pandas->tiledbsoma) (2024.2)\n",
      "Requirement already satisfied: joblib in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from scanpy>=1.9.2->tiledbsoma) (1.4.2)\n",
      "Requirement already satisfied: legacy-api-wrap>=1.4 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from scanpy>=1.9.2->tiledbsoma) (1.4.1)\n",
      "Requirement already satisfied: matplotlib>=3.6 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from scanpy>=1.9.2->tiledbsoma) (3.10.0)\n",
      "Requirement already satisfied: networkx>=2.7 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from scanpy>=1.9.2->tiledbsoma) (3.4.2)\n",
      "Requirement already satisfied: numba>=0.56 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from scanpy>=1.9.2->tiledbsoma) (0.61.0)\n",
      "Requirement already satisfied: patsy!=1.0.0 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from scanpy>=1.9.2->tiledbsoma) (1.0.1)\n",
      "Requirement already satisfied: pynndescent>=0.5 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from scanpy>=1.9.2->tiledbsoma) (0.5.13)\n",
      "Requirement already satisfied: scikit-learn>=1.1 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from scanpy>=1.9.2->tiledbsoma) (1.6.1)\n",
      "Requirement already satisfied: seaborn>=0.13 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from scanpy>=1.9.2->tiledbsoma) (0.13.2)\n",
      "Requirement already satisfied: session-info in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from scanpy>=1.9.2->tiledbsoma) (1.0.0)\n",
      "Requirement already satisfied: statsmodels>=0.13 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from scanpy>=1.9.2->tiledbsoma) (0.14.4)\n",
      "Requirement already satisfied: tqdm in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from scanpy>=1.9.2->tiledbsoma) (4.67.1)\n",
      "Requirement already satisfied: umap-learn!=0.5.0,>=0.5 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from scanpy>=1.9.2->tiledbsoma) (0.5.7)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma) (1.3.1)\n",
      "Requirement already satisfied: cycler>=0.10 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma) (4.55.3)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma) (1.4.8)\n",
      "Requirement already satisfied: pillow>=8 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma) (11.1.0)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma) (3.2.1)\n",
      "Requirement already satisfied: llvmlite<0.45,>=0.44.0dev0 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from numba>=0.56->scanpy>=1.9.2->tiledbsoma) (0.44.0)\n",
      "Requirement already satisfied: six>=1.5 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from python-dateutil>=2.8.2->pandas->tiledbsoma) (1.17.0)\n",
      "Requirement already satisfied: threadpoolctl>=3.1.0 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from scikit-learn>=1.1->scanpy>=1.9.2->tiledbsoma) (3.5.0)\n",
      "Requirement already satisfied: stdlib_list in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from session-info->scanpy>=1.9.2->tiledbsoma) (0.11.0)\n",
      "Requirement already satisfied: colorama in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from tqdm->scanpy>=1.9.2->tiledbsoma) (0.4.6)\n",
      "Using cached somacore-1.0.26-py3-none-any.whl (38 kB)\n",
      "Using cached shapely-2.0.7-cp312-cp312-win_amd64.whl (1.4 MB)\n",
      "Building wheels for collected packages: tiledbsoma\n",
      "  Building wheel for tiledbsoma (pyproject.toml): started\n",
      "  Building wheel for tiledbsoma (pyproject.toml): finished with status 'error'\n",
      "Failed to build tiledbsoma\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  error: subprocess-exited-with-error\n",
      "  \n",
      "  × Building wheel for tiledbsoma (pyproject.toml) did not run successfully.\n",
      "  │ exit code: 1\n",
      "  ╰─> [5 lines of output]\n",
      "      running bdist_wheel\n",
      "      Checking: C:\\Users\\dallo\\AppData\\Local\\Temp\\pip-install-ynm_vlaj\\tiledbsoma_88e4b10649c44357accdba8525d4ecdb\\dist_links\\dist\\lib\\tiledbsoma.dll exists: False\n",
      "      Checking: C:\\Users\\dallo\\AppData\\Local\\Temp\\pip-install-ynm_vlaj\\tiledbsoma_88e4b10649c44357accdba8525d4ecdb\\dist_links\\dist\\bin\\tiledbsoma.dll exists: False\n",
      "      Could not find module 'tiledbsoma.dll' (or one of its dependencies). Try using the full path with constructor syntax.\n",
      "      error: [WinError 2] The system cannot find the file specified\n",
      "      [end of output]\n",
      "  \n",
      "  note: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "  ERROR: Failed building wheel for tiledbsoma\n",
      "\n",
      "[notice] A new release of pip is available: 25.0 -> 25.0.1\n",
      "[notice] To update, run: C:\\Users\\dallo\\AppData\\Local\\Microsoft\\WindowsApps\\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\\python.exe -m pip install --upgrade pip\n",
      "ERROR: Failed to build installable wheels for some pyproject.toml based projects (tiledbsoma)\n"
     ]
    }
   ],
   "source": [
    "!pip install --upgrade  tiledbsoma"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Collecting PyTDC\n",
      "  Downloading pytdc-1.1.14.tar.gz (151 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Collecting accelerate==0.33.0 (from PyTDC)\n",
      "  Downloading accelerate-0.33.0-py3-none-any.whl.metadata (18 kB)\n",
      "Collecting dataclasses<1.0,>=0.6 (from PyTDC)\n",
      "  Downloading dataclasses-0.6-py3-none-any.whl.metadata (3.0 kB)\n",
      "Collecting datasets<2.20.0 (from PyTDC)\n",
      "  Downloading datasets-2.19.2-py3-none-any.whl.metadata (19 kB)\n",
      "Collecting evaluate==0.4.2 (from PyTDC)\n",
      "  Downloading evaluate-0.4.2-py3-none-any.whl.metadata (9.3 kB)\n",
      "Collecting fuzzywuzzy<1.0,>=0.18.0 (from PyTDC)\n",
      "  Downloading fuzzywuzzy-0.18.0-py2.py3-none-any.whl.metadata (4.9 kB)\n",
      "Requirement already satisfied: huggingface_hub<1.0,>=0.20.3 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from PyTDC) (0.27.1)\n",
      "Requirement already satisfied: numpy<2.0.0,>=1.26.4 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from PyTDC) (1.26.4)\n",
      "Collecting openpyxl<4.0.0,>=3.0.10 (from PyTDC)\n",
      "  Downloading openpyxl-3.1.5-py2.py3-none-any.whl.metadata (2.5 kB)\n",
      "Requirement already satisfied: pandas<3.0.0,>=2.1.4 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from PyTDC) (2.2.2)\n",
      "Requirement already satisfied: requests<3.0.0,>=2.31.0 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from PyTDC) (2.32.2)\n",
      "Collecting scikit-learn==1.2.2 (from PyTDC)\n",
      "  Downloading scikit-learn-1.2.2.tar.gz (7.3 MB)\n",
      "     ---------------------------------------- 0.0/7.3 MB ? eta -:--:--\n",
      "     -------------- ------------------------- 2.6/7.3 MB 15.1 MB/s eta 0:00:01\n",
      "     --------------------------------- ------ 6.0/7.3 MB 16.1 MB/s eta 0:00:01\n",
      "     ---------------------------------------- 7.3/7.3 MB 14.9 MB/s eta 0:00:00\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): still running...\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Requirement already satisfied: seaborn<1.0.0,>=0.12.2 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from PyTDC) (0.13.2)\n",
      "Requirement already satisfied: tqdm<5.0.0,>=4.65.0 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from PyTDC) (4.67.1)\n",
      "Collecting transformers==4.43.4 (from PyTDC)\n",
      "  Downloading transformers-4.43.4-py3-none-any.whl.metadata (43 kB)\n",
      "INFO: pip is looking at multiple versions of pytdc to determine which version is compatible with other requirements. This could take a while.\n",
      "Collecting PyTDC\n",
      "  Downloading PyTDC-1.1.13.tar.gz (151 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "  Downloading pytdc-1.1.12.tar.gz (151 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "  Downloading pytdc-1.1.11.tar.gz (151 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "  Downloading pytdc-1.1.10.tar.gz (151 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "  Downloading pytdc-1.1.9.tar.gz (151 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "  Downloading pytdc-1.1.8.tar.gz (151 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "  Downloading pytdc-1.1.7.tar.gz (151 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "INFO: pip is still looking at multiple versions of pytdc to determine which version is compatible with other requirements. This could take a while.\n",
      "  Downloading pytdc-1.1.6.tar.gz (151 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Collecting biopython<2.0,>=1.78 (from PyTDC)\n",
      "  Using cached biopython-1.85-cp312-cp312-win_amd64.whl.metadata (13 kB)\n",
      "Collecting PyTDC\n",
      "  Downloading pytdc-1.1.5.tar.gz (151 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "  Downloading pytdc-1.1.4.tar.gz (151 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Collecting moleculeace==3.0.0 (from PyTDC)\n",
      "  Downloading MoleculeACE-3.0.0-py3-none-any.whl.metadata (549 bytes)\n",
      "Collecting mygene<4.0.0,>=3.2.2 (from PyTDC)\n",
      "  Downloading mygene-3.2.2-py2.py3-none-any.whl.metadata (10 kB)\n",
      "Collecting PyTDC\n",
      "  Downloading pytdc-1.1.3.tar.gz (148 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "  Downloading pytdc-1.1.2.tar.gz (146 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Requirement already satisfied: datasets==2.20.0 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from PyTDC) (2.20.0)\n",
      "INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. See https://pip.pypa.io/warnings/backtracking for guidance. If you want to abort this run, press Ctrl + C.\n",
      "  Downloading pytdc-1.1.1.tar.gz (146 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "  Downloading pytdc-1.1.0.tar.gz (145 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "  Downloading pytdc-1.0.7.tar.gz (144 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "  Downloading pytdc-1.0.6.tar.gz (143 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "  Downloading PyTDC-1.0.0.tar.gz (141 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Collecting cellxgene-census<2.0.0,>=1.10.2 (from PyTDC)\n",
      "  Downloading cellxgene_census-1.16.2-py3-none-any.whl.metadata (5.2 kB)\n",
      "Collecting gget<1.0.0,>=0.28.4 (from PyTDC)\n",
      "  Downloading gget-0.29.0-py3-none-any.whl.metadata (8.0 kB)\n",
      "Requirement already satisfied: pydantic<3.0.0,>=2.6.3 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from PyTDC) (2.10.5)\n",
      "Collecting rdkit<2024.3.1,>=2023.9.5 (from PyTDC)\n",
      "  Downloading rdkit-2023.9.6-cp312-cp312-win_amd64.whl.metadata (4.0 kB)\n",
      "Collecting tiledbsoma<2.0.0,>=1.7.2 (from PyTDC)\n",
      "  Downloading tiledbsoma-1.15.7.tar.gz (685 kB)\n",
      "     ---------------------------------------- 0.0/685.5 kB ? eta -:--:--\n",
      "     -------------------------------------- 685.5/685.5 kB 4.5 MB/s eta 0:00:00\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Collecting yapf<1.0.0,>=0.40.2 (from PyTDC)\n",
      "  Downloading yapf-0.43.0-py3-none-any.whl.metadata (46 kB)\n",
      "Requirement already satisfied: scipy>=1.3.2 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from scikit-learn==1.2.2->PyTDC) (1.13.1)\n",
      "Requirement already satisfied: joblib>=1.1.1 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from scikit-learn==1.2.2->PyTDC) (1.4.2)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from scikit-learn==1.2.2->PyTDC) (3.5.0)\n",
      "Requirement already satisfied: anndata in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from cellxgene-census<2.0.0,>=1.10.2->PyTDC) (0.11.0)\n",
      "Requirement already satisfied: typing-extensions in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from cellxgene-census<2.0.0,>=1.10.2->PyTDC) (4.12.2)\n",
      "Collecting s3fs>=2021.06.1 (from cellxgene-census<2.0.0,>=1.10.2->PyTDC)\n",
      "  Downloading s3fs-2025.2.0-py3-none-any.whl.metadata (1.9 kB)\n",
      "Requirement already satisfied: ipython in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from gget<1.0.0,>=0.28.4->PyTDC) (8.31.0)\n",
      "Requirement already satisfied: matplotlib in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from gget<1.0.0,>=0.28.4->PyTDC) (3.10.0)\n",
      "Collecting mysql-connector-python>=8.0.32 (from gget<1.0.0,>=0.28.4->PyTDC)\n",
      "  Downloading mysql_connector_python-9.2.0-cp312-cp312-win_amd64.whl.metadata (6.2 kB)\n",
      "Requirement already satisfied: beautifulsoup4>=4.10.0 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from gget<1.0.0,>=0.28.4->PyTDC) (4.12.3)\n",
      "Requirement already satisfied: ipywidgets in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from gget<1.0.0,>=0.28.4->PyTDC) (8.1.5)\n",
      "Collecting lxml (from gget<1.0.0,>=0.28.4->PyTDC)\n",
      "  Downloading lxml-5.3.1-cp312-cp312-win_amd64.whl.metadata (3.8 kB)\n",
      "Requirement already satisfied: filelock in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from huggingface_hub<1.0,>=0.20.3->PyTDC) (3.16.1)\n",
      "Requirement already satisfied: fsspec>=2023.5.0 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from huggingface_hub<1.0,>=0.20.3->PyTDC) (2024.5.0)\n",
      "Requirement already satisfied: packaging>=20.9 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from huggingface_hub<1.0,>=0.20.3->PyTDC) (24.2)\n",
      "Requirement already satisfied: pyyaml>=5.1 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from huggingface_hub<1.0,>=0.20.3->PyTDC) (6.0.2)\n",
      "Collecting biothings-client>=0.2.6 (from mygene<4.0.0,>=3.2.2->PyTDC)\n",
      "  Downloading biothings_client-0.4.1-py3-none-any.whl.metadata (10 kB)\n",
      "Collecting et-xmlfile (from openpyxl<4.0.0,>=3.0.10->PyTDC)\n",
      "  Downloading et_xmlfile-2.0.0-py3-none-any.whl.metadata (2.7 kB)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from pandas<3.0.0,>=2.1.4->PyTDC) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from pandas<3.0.0,>=2.1.4->PyTDC) (2024.2)\n",
      "Requirement already satisfied: tzdata>=2022.7 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from pandas<3.0.0,>=2.1.4->PyTDC) (2024.2)\n",
      "Requirement already satisfied: annotated-types>=0.6.0 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from pydantic<3.0.0,>=2.6.3->PyTDC) (0.7.0)\n",
      "Requirement already satisfied: pydantic-core==2.27.2 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from pydantic<3.0.0,>=2.6.3->PyTDC) (2.27.2)\n",
      "Requirement already satisfied: Pillow in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from rdkit<2024.3.1,>=2023.9.5->PyTDC) (11.1.0)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from requests<3.0.0,>=2.31.0->PyTDC) (3.4.1)\n",
      "Requirement already satisfied: idna<4,>=2.5 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from requests<3.0.0,>=2.31.0->PyTDC) (3.10)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from requests<3.0.0,>=2.31.0->PyTDC) (2.3.0)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from requests<3.0.0,>=2.31.0->PyTDC) (2024.12.14)\n",
      "Requirement already satisfied: attrs>=22.2 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from tiledbsoma<2.0.0,>=1.7.2->PyTDC) (24.3.0)\n",
      "Requirement already satisfied: pyarrow in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from tiledbsoma<2.0.0,>=1.7.2->PyTDC) (18.1.0)\n",
      "Requirement already satisfied: scanpy>=1.9.2 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from tiledbsoma<2.0.0,>=1.7.2->PyTDC) (1.10.4)\n",
      "Collecting somacore==1.0.26 (from tiledbsoma<2.0.0,>=1.7.2->PyTDC)\n",
      "  Downloading somacore-1.0.26-py3-none-any.whl.metadata (1.5 kB)\n",
      "Requirement already satisfied: pyarrow-hotfix in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from somacore==1.0.26->tiledbsoma<2.0.0,>=1.7.2->PyTDC) (0.6)\n",
      "Collecting shapely (from somacore==1.0.26->tiledbsoma<2.0.0,>=1.7.2->PyTDC)\n",
      "  Downloading shapely-2.0.7-cp312-cp312-win_amd64.whl.metadata (7.1 kB)\n",
      "Requirement already satisfied: colorama in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from tqdm<5.0.0,>=4.65.0->PyTDC) (0.4.6)\n",
      "Requirement already satisfied: platformdirs>=3.5.1 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from yapf<1.0.0,>=0.40.2->PyTDC) (4.3.6)\n",
      "Requirement already satisfied: array-api-compat!=1.5,>1.4 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from anndata->cellxgene-census<2.0.0,>=1.10.2->PyTDC) (1.10.0)\n",
      "Requirement already satisfied: h5py>=3.6 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from anndata->cellxgene-census<2.0.0,>=1.10.2->PyTDC) (3.12.1)\n",
      "Requirement already satisfied: natsort in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from anndata->cellxgene-census<2.0.0,>=1.10.2->PyTDC) (8.4.0)\n",
      "Requirement already satisfied: soupsieve>1.2 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from beautifulsoup4>=4.10.0->gget<1.0.0,>=0.28.4->PyTDC) (2.6)\n",
      "Requirement already satisfied: httpx>=0.22.0 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from biothings-client>=0.2.6->mygene<4.0.0,>=3.2.2->PyTDC) (0.28.1)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from matplotlib->gget<1.0.0,>=0.28.4->PyTDC) (1.3.1)\n",
      "Requirement already satisfied: cycler>=0.10 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from matplotlib->gget<1.0.0,>=0.28.4->PyTDC) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from matplotlib->gget<1.0.0,>=0.28.4->PyTDC) (4.55.3)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from matplotlib->gget<1.0.0,>=0.28.4->PyTDC) (1.4.8)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from matplotlib->gget<1.0.0,>=0.28.4->PyTDC) (3.2.1)\n",
      "Requirement already satisfied: six>=1.5 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from python-dateutil>=2.8.2->pandas<3.0.0,>=2.1.4->PyTDC) (1.17.0)\n",
      "Collecting aiobotocore<3.0.0,>=2.5.4 (from s3fs>=2021.06.1->cellxgene-census<2.0.0,>=1.10.2->PyTDC)\n",
      "  Downloading aiobotocore-2.19.0-py3-none-any.whl.metadata (23 kB)\n",
      "Collecting fsspec>=2023.5.0 (from huggingface_hub<1.0,>=0.20.3->PyTDC)\n",
      "  Downloading fsspec-2025.2.0-py3-none-any.whl.metadata (11 kB)\n",
      "Requirement already satisfied: aiohttp!=4.0.0a0,!=4.0.0a1 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from s3fs>=2021.06.1->cellxgene-census<2.0.0,>=1.10.2->PyTDC) (3.11.11)\n",
      "Requirement already satisfied: legacy-api-wrap>=1.4 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from scanpy>=1.9.2->tiledbsoma<2.0.0,>=1.7.2->PyTDC) (1.4.1)\n",
      "Requirement already satisfied: networkx>=2.7 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from scanpy>=1.9.2->tiledbsoma<2.0.0,>=1.7.2->PyTDC) (3.4.2)\n",
      "Requirement already satisfied: numba>=0.56 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from scanpy>=1.9.2->tiledbsoma<2.0.0,>=1.7.2->PyTDC) (0.61.0)\n",
      "Requirement already satisfied: patsy!=1.0.0 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from scanpy>=1.9.2->tiledbsoma<2.0.0,>=1.7.2->PyTDC) (1.0.1)\n",
      "Requirement already satisfied: pynndescent>=0.5 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from scanpy>=1.9.2->tiledbsoma<2.0.0,>=1.7.2->PyTDC) (0.5.13)\n",
      "Requirement already satisfied: session-info in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from scanpy>=1.9.2->tiledbsoma<2.0.0,>=1.7.2->PyTDC) (1.0.0)\n",
      "Requirement already satisfied: statsmodels>=0.13 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from scanpy>=1.9.2->tiledbsoma<2.0.0,>=1.7.2->PyTDC) (0.14.4)\n",
      "Requirement already satisfied: umap-learn!=0.5.0,>=0.5 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from scanpy>=1.9.2->tiledbsoma<2.0.0,>=1.7.2->PyTDC) (0.5.7)\n",
      "Requirement already satisfied: decorator in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from ipython->gget<1.0.0,>=0.28.4->PyTDC) (5.1.1)\n",
      "Requirement already satisfied: jedi>=0.16 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from ipython->gget<1.0.0,>=0.28.4->PyTDC) (0.19.2)\n",
      "Requirement already satisfied: matplotlib-inline in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from ipython->gget<1.0.0,>=0.28.4->PyTDC) (0.1.7)\n",
      "Requirement already satisfied: prompt_toolkit<3.1.0,>=3.0.41 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from ipython->gget<1.0.0,>=0.28.4->PyTDC) (3.0.48)\n",
      "Requirement already satisfied: pygments>=2.4.0 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from ipython->gget<1.0.0,>=0.28.4->PyTDC) (2.19.1)\n",
      "Requirement already satisfied: stack_data in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from ipython->gget<1.0.0,>=0.28.4->PyTDC) (0.6.3)\n",
      "Requirement already satisfied: traitlets>=5.13.0 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from ipython->gget<1.0.0,>=0.28.4->PyTDC) (5.14.3)\n",
      "Requirement already satisfied: comm>=0.1.3 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from ipywidgets->gget<1.0.0,>=0.28.4->PyTDC) (0.2.2)\n",
      "Requirement already satisfied: widgetsnbextension~=4.0.12 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from ipywidgets->gget<1.0.0,>=0.28.4->PyTDC) (4.0.13)\n",
      "Requirement already satisfied: jupyterlab-widgets~=3.0.12 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from ipywidgets->gget<1.0.0,>=0.28.4->PyTDC) (3.0.13)\n",
      "Collecting aioitertools<1.0.0,>=0.5.1 (from aiobotocore<3.0.0,>=2.5.4->s3fs>=2021.06.1->cellxgene-census<2.0.0,>=1.10.2->PyTDC)\n",
      "  Downloading aioitertools-0.12.0-py3-none-any.whl.metadata (3.8 kB)\n",
      "Collecting botocore<1.36.4,>=1.36.0 (from aiobotocore<3.0.0,>=2.5.4->s3fs>=2021.06.1->cellxgene-census<2.0.0,>=1.10.2->PyTDC)\n",
      "  Downloading botocore-1.36.3-py3-none-any.whl.metadata (5.7 kB)\n",
      "Requirement already satisfied: jmespath<2.0.0,>=0.7.1 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from aiobotocore<3.0.0,>=2.5.4->s3fs>=2021.06.1->cellxgene-census<2.0.0,>=1.10.2->PyTDC) (1.0.1)\n",
      "Requirement already satisfied: multidict<7.0.0,>=6.0.0 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from aiobotocore<3.0.0,>=2.5.4->s3fs>=2021.06.1->cellxgene-census<2.0.0,>=1.10.2->PyTDC) (6.1.0)\n",
      "Requirement already satisfied: wrapt<2.0.0,>=1.10.10 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from aiobotocore<3.0.0,>=2.5.4->s3fs>=2021.06.1->cellxgene-census<2.0.0,>=1.10.2->PyTDC) (1.17.2)\n",
      "Requirement already satisfied: aiohappyeyeballs>=2.3.0 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census<2.0.0,>=1.10.2->PyTDC) (2.4.4)\n",
      "Requirement already satisfied: aiosignal>=1.1.2 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census<2.0.0,>=1.10.2->PyTDC) (1.3.2)\n",
      "Requirement already satisfied: frozenlist>=1.1.1 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census<2.0.0,>=1.10.2->PyTDC) (1.5.0)\n",
      "Requirement already satisfied: propcache>=0.2.0 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census<2.0.0,>=1.10.2->PyTDC) (0.2.1)\n",
      "Requirement already satisfied: yarl<2.0,>=1.17.0 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census<2.0.0,>=1.10.2->PyTDC) (1.18.3)\n",
      "Requirement already satisfied: anyio in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from httpx>=0.22.0->biothings-client>=0.2.6->mygene<4.0.0,>=3.2.2->PyTDC) (4.8.0)\n",
      "Requirement already satisfied: httpcore==1.* in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from httpx>=0.22.0->biothings-client>=0.2.6->mygene<4.0.0,>=3.2.2->PyTDC) (1.0.7)\n",
      "Requirement already satisfied: h11<0.15,>=0.13 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from httpcore==1.*->httpx>=0.22.0->biothings-client>=0.2.6->mygene<4.0.0,>=3.2.2->PyTDC) (0.14.0)\n",
      "Requirement already satisfied: parso<0.9.0,>=0.8.4 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from jedi>=0.16->ipython->gget<1.0.0,>=0.28.4->PyTDC) (0.8.4)\n",
      "Requirement already satisfied: llvmlite<0.45,>=0.44.0dev0 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from numba>=0.56->scanpy>=1.9.2->tiledbsoma<2.0.0,>=1.7.2->PyTDC) (0.44.0)\n",
      "Requirement already satisfied: wcwidth in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from prompt_toolkit<3.1.0,>=3.0.41->ipython->gget<1.0.0,>=0.28.4->PyTDC) (0.2.13)\n",
      "Requirement already satisfied: stdlib_list in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from session-info->scanpy>=1.9.2->tiledbsoma<2.0.0,>=1.7.2->PyTDC) (0.11.0)\n",
      "Requirement already satisfied: executing>=1.2.0 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from stack_data->ipython->gget<1.0.0,>=0.28.4->PyTDC) (2.1.0)\n",
      "Requirement already satisfied: asttokens>=2.1.0 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from stack_data->ipython->gget<1.0.0,>=0.28.4->PyTDC) (3.0.0)\n",
      "Requirement already satisfied: pure-eval in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from stack_data->ipython->gget<1.0.0,>=0.28.4->PyTDC) (0.2.3)\n",
      "Requirement already satisfied: sniffio>=1.1 in c:\\users\\dallo\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from anyio->httpx>=0.22.0->biothings-client>=0.2.6->mygene<4.0.0,>=3.2.2->PyTDC) (1.3.1)\n",
      "Downloading biopython-1.85-cp312-cp312-win_amd64.whl (2.8 MB)\n",
      "   ---------------------------------------- 0.0/2.8 MB ? eta -:--:--\n",
      "   ----------- ---------------------------- 0.8/2.8 MB 4.2 MB/s eta 0:00:01\n",
      "   -------------------------- ------------- 1.8/2.8 MB 4.8 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 2.8/2.8 MB 4.5 MB/s eta 0:00:00\n",
      "Downloading cellxgene_census-1.16.2-py3-none-any.whl (54 kB)\n",
      "Downloading dataclasses-0.6-py3-none-any.whl (14 kB)\n",
      "Downloading fuzzywuzzy-0.18.0-py2.py3-none-any.whl (18 kB)\n",
      "Downloading gget-0.29.0-py3-none-any.whl (43.1 MB)\n",
      "   ---------------------------------------- 0.0/43.1 MB ? eta -:--:--\n",
      "    --------------------------------------- 1.0/43.1 MB 5.6 MB/s eta 0:00:08\n",
      "   - -------------------------------------- 2.1/43.1 MB 5.1 MB/s eta 0:00:09\n",
      "   --- ------------------------------------ 3.4/43.1 MB 5.3 MB/s eta 0:00:08\n",
      "   ---- ----------------------------------- 4.7/43.1 MB 5.7 MB/s eta 0:00:07\n",
      "   ----- ---------------------------------- 6.0/43.1 MB 5.8 MB/s eta 0:00:07\n",
      "   ------ --------------------------------- 7.1/43.1 MB 5.7 MB/s eta 0:00:07\n",
      "   -------- ------------------------------- 8.7/43.1 MB 6.0 MB/s eta 0:00:06\n",
      "   --------- ------------------------------ 10.0/43.1 MB 6.0 MB/s eta 0:00:06\n",
      "   ---------- ----------------------------- 11.3/43.1 MB 6.0 MB/s eta 0:00:06\n",
      "   ------------ --------------------------- 13.1/43.1 MB 6.2 MB/s eta 0:00:05\n",
      "   ------------- -------------------------- 14.7/43.1 MB 6.3 MB/s eta 0:00:05\n",
      "   -------------- ------------------------- 16.0/43.1 MB 6.3 MB/s eta 0:00:05\n",
      "   ---------------- ----------------------- 17.6/43.1 MB 6.4 MB/s eta 0:00:04\n",
      "   ----------------- ---------------------- 19.1/43.1 MB 6.5 MB/s eta 0:00:04\n",
      "   ------------------ --------------------- 20.4/43.1 MB 6.5 MB/s eta 0:00:04\n",
      "   -------------------- ------------------- 22.3/43.1 MB 6.6 MB/s eta 0:00:04\n",
      "   ---------------------- ----------------- 24.1/43.1 MB 6.7 MB/s eta 0:00:03\n",
      "   ------------------------ --------------- 26.0/43.1 MB 6.8 MB/s eta 0:00:03\n",
      "   ------------------------- -------------- 27.5/43.1 MB 6.9 MB/s eta 0:00:03\n",
      "   --------------------------- ------------ 29.6/43.1 MB 7.0 MB/s eta 0:00:02\n",
      "   ----------------------------- ---------- 31.5/43.1 MB 7.1 MB/s eta 0:00:02\n",
      "   ------------------------------- -------- 33.8/43.1 MB 7.3 MB/s eta 0:00:02\n",
      "   --------------------------------- ------ 35.9/43.1 MB 7.4 MB/s eta 0:00:01\n",
      "   ----------------------------------- ---- 38.3/43.1 MB 7.5 MB/s eta 0:00:01\n",
      "   ------------------------------------- -- 40.6/43.1 MB 7.7 MB/s eta 0:00:01\n",
      "   ---------------------------------------  42.7/43.1 MB 7.7 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 43.1/43.1 MB 7.5 MB/s eta 0:00:00\n",
      "Downloading mygene-3.2.2-py2.py3-none-any.whl (5.4 kB)\n",
      "Downloading openpyxl-3.1.5-py2.py3-none-any.whl (250 kB)\n",
      "Downloading rdkit-2023.9.6-cp312-cp312-win_amd64.whl (21.0 MB)\n",
      "   ---------------------------------------- 0.0/21.0 MB ? eta -:--:--\n",
      "   --- ------------------------------------ 1.8/21.0 MB 9.1 MB/s eta 0:00:03\n",
      "   ------ --------------------------------- 3.4/21.0 MB 8.4 MB/s eta 0:00:03\n",
      "   ---------- ----------------------------- 5.5/21.0 MB 8.8 MB/s eta 0:00:02\n",
      "   ------------- -------------------------- 7.3/21.0 MB 8.9 MB/s eta 0:00:02\n",
      "   ------------------ --------------------- 9.7/21.0 MB 9.1 MB/s eta 0:00:02\n",
      "   --------------------- ------------------ 11.5/21.0 MB 9.4 MB/s eta 0:00:02\n",
      "   -------------------------- ------------- 14.2/21.0 MB 9.6 MB/s eta 0:00:01\n",
      "   ------------------------------- -------- 16.5/21.0 MB 9.8 MB/s eta 0:00:01\n",
      "   ----------------------------------- ---- 18.9/21.0 MB 9.8 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 20.4/21.0 MB 9.6 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 21.0/21.0 MB 9.3 MB/s eta 0:00:00\n",
      "Downloading somacore-1.0.26-py3-none-any.whl (38 kB)\n",
      "Downloading yapf-0.43.0-py3-none-any.whl (256 kB)\n",
      "Downloading biothings_client-0.4.1-py3-none-any.whl (46 kB)\n",
      "Downloading mysql_connector_python-9.2.0-cp312-cp312-win_amd64.whl (16.1 MB)\n",
      "   ---------------------------------------- 0.0/16.1 MB ? eta -:--:--\n",
      "   ----- ---------------------------------- 2.1/16.1 MB 10.7 MB/s eta 0:00:02\n",
      "   --------- ------------------------------ 3.9/16.1 MB 9.8 MB/s eta 0:00:02\n",
      "   -------------- ------------------------- 5.8/16.1 MB 9.5 MB/s eta 0:00:02\n",
      "   -------------------- ------------------- 8.1/16.1 MB 9.7 MB/s eta 0:00:01\n",
      "   -------------------------- ------------- 10.5/16.1 MB 10.1 MB/s eta 0:00:01\n",
      "   ------------------------------- -------- 12.8/16.1 MB 10.3 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 15.5/16.1 MB 10.5 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 16.1/16.1 MB 10.0 MB/s eta 0:00:00\n",
      "Downloading s3fs-2025.2.0-py3-none-any.whl (30 kB)\n",
      "Downloading fsspec-2025.2.0-py3-none-any.whl (184 kB)\n",
      "Downloading et_xmlfile-2.0.0-py3-none-any.whl (18 kB)\n",
      "Downloading lxml-5.3.1-cp312-cp312-win_amd64.whl (3.8 MB)\n",
      "   ---------------------------------------- 0.0/3.8 MB ? eta -:--:--\n",
      "   ------------------------ --------------- 2.4/3.8 MB 11.2 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 3.8/3.8 MB 10.8 MB/s eta 0:00:00\n",
      "Downloading aiobotocore-2.19.0-py3-none-any.whl (77 kB)\n",
      "Downloading shapely-2.0.7-cp312-cp312-win_amd64.whl (1.4 MB)\n",
      "   ---------------------------------------- 0.0/1.4 MB ? eta -:--:--\n",
      "   ---------------------------------------- 1.4/1.4 MB 10.8 MB/s eta 0:00:00\n",
      "Downloading aioitertools-0.12.0-py3-none-any.whl (24 kB)\n",
      "Downloading botocore-1.36.3-py3-none-any.whl (13.3 MB)\n",
      "   ---------------------------------------- 0.0/13.3 MB ? eta -:--:--\n",
      "   ------- -------------------------------- 2.4/13.3 MB 11.2 MB/s eta 0:00:01\n",
      "   -------------- ------------------------- 4.7/13.3 MB 11.0 MB/s eta 0:00:01\n",
      "   --------------------- ------------------ 7.1/13.3 MB 10.9 MB/s eta 0:00:01\n",
      "   -------------------------- ------------- 8.9/13.3 MB 10.4 MB/s eta 0:00:01\n",
      "   ---------------------------------- ----- 11.5/13.3 MB 10.8 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 13.3/13.3 MB 10.4 MB/s eta 0:00:00\n",
      "Building wheels for collected packages: PyTDC, scikit-learn, tiledbsoma\n",
      "  Building wheel for PyTDC (pyproject.toml): started\n",
      "  Building wheel for PyTDC (pyproject.toml): finished with status 'done'\n",
      "  Created wheel for PyTDC: filename=PyTDC-1.0.0-py3-none-any.whl size=178032 sha256=e975cb53463e1c85a1892507a3d78523b2112833c56174b277319bbc469ce132\n",
      "  Stored in directory: c:\\users\\dallo\\appdata\\local\\pip\\cache\\wheels\\c8\\b2\\df\\def482293314b03bd5c2d72c94088b79e70437b4a731099f83\n",
      "  Building wheel for scikit-learn (pyproject.toml): started\n",
      "  Building wheel for scikit-learn (pyproject.toml): still running...\n",
      "  Building wheel for scikit-learn (pyproject.toml): still running...\n",
      "  Building wheel for scikit-learn (pyproject.toml): finished with status 'done'\n",
      "  Created wheel for scikit-learn: filename=scikit_learn-1.2.2-cp312-cp312-win_amd64.whl size=9009556 sha256=903310f6751206cc012c1143cbc6307a5cff3180cae9c90b76960410fbb96028\n",
      "  Stored in directory: c:\\users\\dallo\\appdata\\local\\pip\\cache\\wheels\\24\\f8\\77\\ae90c181b806f450a6fec8c8f794594e7c92fa79d7ca27e656\n",
      "  Building wheel for tiledbsoma (pyproject.toml): started\n",
      "  Building wheel for tiledbsoma (pyproject.toml): finished with status 'error'\n",
      "Successfully built PyTDC scikit-learn\n",
      "Failed to build tiledbsoma\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  error: subprocess-exited-with-error\n",
      "  \n",
      "  × Building wheel for tiledbsoma (pyproject.toml) did not run successfully.\n",
      "  │ exit code: 1\n",
      "  ╰─> [5 lines of output]\n",
      "      running bdist_wheel\n",
      "      Checking: C:\\Users\\dallo\\AppData\\Local\\Temp\\pip-install-cozh1k4k\\tiledbsoma_5e1003847ae9490f96848b40606cf79d\\dist_links\\dist\\lib\\tiledbsoma.dll exists: False\n",
      "      Checking: C:\\Users\\dallo\\AppData\\Local\\Temp\\pip-install-cozh1k4k\\tiledbsoma_5e1003847ae9490f96848b40606cf79d\\dist_links\\dist\\bin\\tiledbsoma.dll exists: False\n",
      "      Could not find module 'tiledbsoma.dll' (or one of its dependencies). Try using the full path with constructor syntax.\n",
      "      error: [WinError 2] The system cannot find the file specified\n",
      "      [end of output]\n",
      "  \n",
      "  note: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "  ERROR: Failed building wheel for tiledbsoma\n",
      "\n",
      "[notice] A new release of pip is available: 25.0 -> 25.0.1\n",
      "[notice] To update, run: C:\\Users\\dallo\\AppData\\Local\\Microsoft\\WindowsApps\\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\\python.exe -m pip install --upgrade pip\n",
      "ERROR: Failed to build installable wheels for some pyproject.toml based projects (tiledbsoma)\n"
     ]
    }
   ],
   "source": [
    "!pip install PyTDC            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'tdc'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[7], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mtdc\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmulti_pred\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01manndata_dataset\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m DataLoader\n\u001b[0;32m      2\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mtdc\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m tdc_hf_interface\n\u001b[0;32m      3\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mtdc\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmodel_server\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtokenizers\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mscgpt\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m scGPTTokenizer\n",
      "\u001b[1;31mModuleNotFoundError\u001b[0m: No module named 'tdc'"
     ]
    }
   ],
   "source": [
    "from tdc.multi_pred.anndata_dataset import DataLoader\n",
    "from tdc import tdc_hf_interface\n",
    "from tdc.model_server.tokenizers.scgpt import scGPTTokenizer\n",
    "import torch\n",
    "\n",
    "# an example dataset\n",
    "adata = DataLoader(\"cellxgene_sample_small\",\n",
    "                   \"./data\",\n",
    "                   dataset_names=[\"cellxgene_sample_small\"],\n",
    "                   no_convert=True).adata\n",
    "\n",
    "# code for loading the model and performing inference\n",
    "scgpt = tdc_hf_interface(\"scGPT\")\n",
    "model = scgpt.load()  # This line can cause segmentation fault on inappropriate setup\n",
    "tokenizer = scGPTTokenizer()\n",
    "gene_ids = adata.var[\"feature_name\"].to_numpy(\n",
    ")  # Convert to numpy array\n",
    "tokenized_data = tokenizer.tokenize_cell_vectors(\n",
    "    adata.X.toarray(), gene_ids)\n",
    "mask = torch.tensor([x != 0 for x in tokenized_data[0][1]],\n",
    "                    dtype=torch.bool)\n",
    "\n",
    "# Extract first embedding\n",
    "first_embed = model(tokenized_data[0][0],\n",
    "                    tokenized_data[0][1],\n",
    "                    attention_mask=mask)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
